docs_src/design/ucr/0002-Device-Parent-Child-Relationships.md

```diff
@@ -0,0 +1,115 @@
+where it is necessary to know which devices are parents and which are their children.
+
+These considerations are most important for gateways that are implemented with the EdgeX framework,
+since there are potentially very many south-bound devices connected to a system.
+
+Examples are
+* BACnet - where only one "main" device is present at the point of external connection (eg, UDP port 0xBAC0) and all other devices must be presented as "virtually routed devices" connected to that main "virtual router" device.
+* Azure IoT Hub - where the normal connection for IoT Plug and Play / Digital Twin is for a single device, and any other devices need to somehow fall under that device (eg, with Device Twin "Modules")
+* UI device presentation - where child devices are grouped under their parent, often rolled up until they are expanded to show their data
+* Multi-tenant deployments of multi-point energy meters - where a main meter has up to 80 Branch Circuit Monitoring (BCM) points connected to it, each BCM modeled as a Device consisting of the same 6 or so energy channels (Device Resources), and each BCM is assigned to a particular tenant. Tenants will be given access to the data from their BCM point(s) but not those of other tenants. A gateway may connect more than one of these multi-point energy meters.
+
+Since there are multiple similar uses for this relationship information on the north side, it is proposed to locate
+this relationship metadata in the Device object as accessed from core-metadata by all services, rather than to 
+locate it in each north-bound service (which would be particularly problematic for the UI, which gets its data through REST APIs).
+
+The sound-bound Device Service that creates a Device is ideally the service which establishes this relationship data, though it is possible that it is unaware of the parent-child relationship. It should be permitted, therefore, for this relationship information to also be set by north-bound services (most likely the UI) and simply ignored by the south-bound Device Service.
+
+It is also necessary to indicate which device is the "main" or "publisher" device (ie, the gateway device), 
+as any devices without a configured relationship can be inferred to be children of that device.
+
+It is frequently a pattern in data servers to "walk the device tree", starting with the main device, then 
+recursively processing its direct child devices, and then the child devices (if any) of those devices, until
+all devices have been processed. This is normally part of the initialization of device data for a server,
+since the parent must be processed and initialized before its child devices. Consequently, there is a need
+for a means to answer the question "What are the child devices (if any) of device x.y.z?"; this is commonly done
+either with the device structure listing its children, or by providing a query that can answer this question.
```

Comment:
This UCR was discussed a bit in both the Device Service WG and Core WG meetings this week (8/8 and 8/9).  Of concern was the requirements against the core command and device services.  If we had a parent/child relationship, what would the requirement be for commands sent to a parent device?  Would the command propagate to the children?  If so, what happens when one of the children had an error?  Would all the commands be somehow rolled back?  How would results be returned (for something like a set)?  Would it be an array of results?  Does the command actually get done on the parent device as well or is it just a propagator?

Same can be said of calling on commands directly of the device service since command actually just calls on the device service APIs.

Would the device profile apply in the same way to parent and child or is there a need to describe the resources and properties of a parent differently from the child?

When removing a parent, are the associated children also removed or do they become orphan devices?

We don't have to get into the specifics of implementation yet, but as it pertains to requirements, what are the needs of other systems to make commands on these parent/children relationships?

Reply:
> This UCR was discussed a bit in both the Device Service WG and Core WG meetings this week (8/8 and 8/9). Of concern was the requirements against the core command and device services. If we had a parent/child relationship, what would the requirement be for commands sent to a parent device? Would the command propagate to the children? If so, what happens when one of the children had an error? Would all the commands be somehow rolled back? How would results be returned (for something like a set)? Would it be an array of results? Does the command actually get done on the parent device as well or is it just a propagator?

I did not expect this to affect commands. The UCR appears to focus on metadata but I just noticed sudden mentions of core-command and core-data in the discussion section. IMO, the UCR needs major rework if core-command and core-data services need to support this relationship.

Reply:
Per our discussion, I've revised the comments to make it clear that we are not expecting EdgeX core services to do anything with this "information" relationship data.